### PR TITLE
opt: adds partialidx package for testing predicate implication

### PR DESCRIPTION
**Pull request by mgartner**
This commit adds a new package, `partialidx`, and a function
`FiltersImplyPredicate` which is responsible for determining if a
`FiltersExpr` implies a partial index's predicate `ScalarExpr`.

For now, the logic for proving implication is very simple. It can only
prove implication when a single `FiltersItem` is an exact match to the
predicate. In predicate implication nomenclature, this means that it
only supports `atom => atom` when the atoms are identical and `AND-expr
=> atom` when the `AND-expr` is at the root of the filters and one of
its children is identical to the atom. (An atom is any expression that
is not an AND or OR).

Release note: None

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm: 

> Reviewed 5 of 5 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-:-M9V6flR7DQ20gpkAx79:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/partialidx/predicate.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9V5Jf90plBvbGd0aOK:-M9V5JfA9fryfxvwAa_V:bv6fe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate.go#L20)):*
> ```Go
> // FiltersImplyPredicate attempts to prove that a partial index predicate is
> // implied by the given filters. If implication is proven, the function returns
> // a simplified filters expression and true. If implication cannot be proven,
> ```

This text implies that the simplified filters are equivalent to the original filters, but in the code below you're actually removing the predicate. I think you just need to clarify that it returns the remaining filters not including the predicate (and give a bit of intuition of why this is desirable).
___
*[pkg/sql/opt/partialidx/predicate.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9UzgPF4BEFELYuRGAW:-M9UzgPG-tm_L36YsYhI:bk403p8) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate.go#L30)):*
> ```Go
> // The logic for testing implication of ScalarExpr with conjunctions,
> // disjunctions, and atoms (anything that is not an AND or OR) is directly
> // influenced by Postgres's predtest library.
> ```

[nit] "directly influenced" sounds kind of weird. I would just say "is the same as", or explain what the differences are.
___
*[pkg/sql/opt/partialidx/predicate_test.go, line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9V5gfZ2SIHXbC3xcaF:-M9V5gfZ2SIHXbC3xcaG:b-ge37vw) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate_test.go#L46)):*
> ```Go
> //      [filter expression]
> //      =>
> //      [prediction expression]
> ```

prediction -> predicate

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50034)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/partialidx/predicate.go, line 20 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9V5Jf90plBvbGd0aOK:-M9VhBs_7TiKs1t_r6lS:bkr31f5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate.go#L20)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

This text implies that the simplified filters are equivalent to the original filters, but in the code below you're actually removing the predicate. I think you just need to clarify that it returns the remaining filters not including the predicate (and give a bit of intuition of why this is desirable).
</blockquote></details>

I had a blurb about this in the paragraph directly below, but I can see how this is confusing. I gave it another try, but TBH I'm struggling to write this out clearly. I very much appreciate any further feedback!
___
*[pkg/sql/opt/partialidx/predicate.go, line 30 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9UzgPF4BEFELYuRGAW:-M9VhWRs9L8hO1ubC-14:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate.go#L30)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] "directly influenced" sounds kind of weird. I would just say "is the same as", or explain what the differences are.
</blockquote></details>

Done.
___
*[pkg/sql/opt/partialidx/predicate_test.go, line 46 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9V5gfZ2SIHXbC3xcaF:-M9Vhiye3Sv6399i6IoH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0945304c3d6475b27dcec452372d71ae6275cd41/pkg/sql/opt/partialidx/predicate_test.go#L46)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

prediction -> predicate
</blockquote></details>

Done.

**Review by rytaft — APPROVED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50034)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @mgartner)

___
*[pkg/sql/opt/partialidx/predicate.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9YQeHBCED_hpKo_nPg:-M9YQeHBCED_hpKo_nPh:bawrjf1) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d398bc685e9c338f025a2f39c134972c2f45ba2/pkg/sql/opt/partialidx/predicate.go#L29)):*
> ```Go
> // matches the predicate expression are removed. This reduces the complexity of
> // the filters, allow any columns that are referenced only in the filters to be
> // pruned from the query plan.
> ```

I'd rephrase to something like this:

"The remaining filters are identical to the input filters except that any parts that exactly match the predicate expression are removed. When the remaining filters are applied to a partial index with the given predicate, the resulting expression is equivalent to the original expression. This reduces the complexity of the filters and allows any columns that are referenced only in the filters to be pruned from the query plan."

**Review by mgartner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50034)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @rytaft)

___
*[pkg/sql/opt/partialidx/predicate.go, line 29 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/50034#-M9YQeHBCED_hpKo_nPg:-M9_6Kuk2wMXHBIeXMWD:b-prndz5) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d398bc685e9c338f025a2f39c134972c2f45ba2/pkg/sql/opt/partialidx/predicate.go#L29)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I'd rephrase to something like this:

"The remaining filters are identical to the input filters except that any parts that exactly match the predicate expression are removed. When the remaining filters are applied to a partial index with the given predicate, the resulting expression is equivalent to the original expression. This reduces the complexity of the filters and allows any columns that are referenced only in the filters to be pruned from the query plan."

</blockquote></details>

Done. Thanks!!!

**Review by rytaft — APPROVED** [review]
> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/50034)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)
